rfcs/006-package-management.md

```diff
@@ -0,0 +1,243 @@
+another version without complaining. If nothing has changed since the lock-file
+was created, it should always resolve the same versions.
+
+## Version compatibility and resolution
+
+How do we handle a package that gets imported multiple times in the dependency
+tree?
+
+For path and git dependencies, there isn't much choice. Dependencies
+from the registry are the most interesting. Fortunately, there are fairly
+well-established conventions for specifying ranges of versions (like ">=1.0
+<3.0", or "^1.2"). What's less clear is how to handle multiple packages with
+overlapping ranges. Some languages (e.g. python) insist that each package
+resolves to a single version across the whole dependency tree. Other languages
+allow multiple versions, keeping track of which package in the dependency tree
+needs to import which version of a package.
+
+I think we want to allow multiple versions of a package; the alternative can be
+fragile and annoying. But then we need to figure out how many different versions
+to allow. There's a trade-off: if we allow pulling in a different version
+every time a package gets imported, solving the dependency graph is easy.
+But it increases the chance of getting incompatibilities at runtime: we might
+accidentally get a value from `util@1.1` and try to pass it to an incompatible
+function defined in `util@1.2`. Pulling in too many different versions also
+increases the total number of packages in the dependency graph.
+
+The current prototype uses a strategy similar to cargo: it divides package
+versions into semver-delimited "bins" and allows resolution to choose at most
+one version from each bin. That is, we can have a `util@2.2` and a `util@1.2` in
+the same dependency tree, but not a `util@1.2` and a `util@1.1`.
+
+## The registry
+
+How should we manage the global registry? There's a potential for incurring
+substantial maintenance costs here, so we should be careful.
+
+We will provide a git repo, at a hard-coded location, to serve as the registry.
+This repo will contain the "index", but not the actual package contents. It
+will contain one file per package, each of which contains a line per version.
+Each entry specifies the location of the package (currently required to be on
+github) and its git tree hash. This ensures that packages are immutable, but it
+doesn't stop them from disappearing: we don't keep a copy of the actual package
+contents.
+
+The registry entries are named like "github/\<org\>/\<package\>" (where in the
+future we might support places other than github). This allows the package
+registry to automatically discover new package versions: to find the latest
+versions of "github/tweag/json-schema-lib", we simply fetch the repository at
+`github.com/tweag/json-schema-lib` and look for tags that look like version
+numbers. Initially, we will scrape packages daily in a cron job. Eventually we
+will allow people to automatically request re-scrapes of specific packages.
```

Comment:
I'm not sure to understand exactly when this process takes place. Is it to ease the addition of a new package in the registry (automatic discovery)? Or is it more than that, like re-generating the index with auto-updates for existing pacakges? In the latter case, isn't that a bit fragile if it's entirely automated but based on heuristics?

Reply:
I was thinking initially of manual package addition + automatic discovery of new versions. The tag-scanning part might be too fragile, I agree. The only alternative I can think of would require a bit more effort but it might actually not be too bad: instead of running a cron job, have a service where you can POST to `nickel-mine.org/release/github/tweag/json-schema-lib?v=0.1.0&rev=df1234`, and if the specified repo has the '0.1.0' tag pointing at revision df1234 it will publish that version.

Reply:
Initial version could use pull requests to the index for updating.